### ISSUE TITLE
Want better error on refusal to snapshot disk attached to stopped instance

### ISSUE THREAD

**Issue by david-crespo** [oxidecomputer/omicron#5108]
I tried to create a snapshot of a disk attached to a stopped instance. This is not allowed, but the 503 response surprised me — I would expect a 400 with a more helpful message.

```json
{
  "request_id": "f14e26d3-f122-4f12-b284-daf99c2386c3",
  "error_code": "ServiceNotAvailable",
  "message": "Service Unavailable"
}
```

I believe the error comes from here.

https://github.com/oxidecomputer/omicron/blob/915276d47096a9ce076453eb75aaa9de29477e80/nexus/src/app/sagas/snapshot_create.rs#L841-L851

From the Nexus log:

```
error_message_external = Service Unavailable
error_message_internal = saga ACTION error at node "disk_generation_number": disk is in state DiskState(Attached(33e6d4a4-6cc7-440f-8a9f-152d94136191))
```

Something odd might be going on, though, because in theory we are checking in the instance state before we even run the saga and giving precisely the 400 I would expect. Or maybe the problem is that we're hitting the `None` arm of the match and therefore not getting an error? I'll need to look into this further.

https://github.com/oxidecomputer/omicron/blob/915276d47096a9ce076453eb75aaa9de29477e80/nexus/src/app/snapshot.rs#L110-L129

**Comment by smklein** [comment]
FYI @jmpesp - isn't this the sorta scenario the pantry was designed to solve? Or is that only for disks that are explicitly not attached to **any** instance?

**Comment by david-crespo** [comment]
https://github.com/oxidecomputer/omicron/blob/915276d47096a9ce076453eb75aaa9de29477e80/nexus/src/app/sagas/snapshot_create.rs#L254

**Comment by jmpesp** [comment]
> FYI @jmpesp - isn't this the sorta scenario the pantry was designed to solve? Or is that only for disks that are explicitly not attached to **any** instance?

It is one of them yes! Currently the pantry is used for exactly this when the disk is not attached to any instance, and it'd have to be used to implement this as well.

Where the instance is stopped but the disk is attached, the difficulty lies in that snapshot request racing with a "start instance" request. It's not impossible to work out a way to do this! I haven't figured out a satisfying way yet though.

See also https://github.com/oxidecomputer/omicron/issues/3289

**Comment by david-crespo** [comment]
Why is this check not getting tripped? Does a stopped instance not have a vmm?

https://github.com/oxidecomputer/omicron/blob/915276d47096a9ce076453eb75aaa9de29477e80/nexus/src/app/snapshot.rs#L110-L129

**Comment by augustuswm** [comment]
That is correct. The vmm record is tied to the running lifecycle of an instance. Once an instance is stopped the vmm record is torn down. We had a user on the colo rack run in to this just today. It seems like we want an additional check in the meantime such that if a disk is attached and it does not have a vmm record that we bail out early.
